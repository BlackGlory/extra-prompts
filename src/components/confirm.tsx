import { Box, Text } from 'ink'
import { TextInput } from '@inkjs/ui'

interface IConfirmProps {
  message: string
  defaultValue: boolean

  onSubmit(value: boolean): void
}

export function Confirm(props: IConfirmProps) {
  const { message, defaultValue, onSubmit } = props

  return (
    <Box>
      <Text>{message}{defaultValue ? '(Y/n)' : '(y/N)'}: </Text>
      <TextInput
        onSubmit={value => {
          const lowerCaseValue = value.toLowerCase()
          if (lowerCaseValue.startsWith('y')) {
            onSubmit(true)
          } else if (lowerCaseValue.startsWith('n')) {
            onSubmit(false)
          } else {
            onSubmit(defaultValue)
          }
        }} 
      />
    </Box>
  )
}
