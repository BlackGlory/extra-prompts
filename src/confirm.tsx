import { Confirm } from '@components/confirm.js'
import { render } from 'ink'

export function confirm(message: string, defaultValue: boolean): Promise<boolean> {
  return new Promise(resolve => {
    const { unmount } = render(
      <Confirm
        defaultValue={defaultValue}
        message={message}
        onSubmit={value => {
          unmount()
          resolve(value)
        }}
      />
    )
  })
}
