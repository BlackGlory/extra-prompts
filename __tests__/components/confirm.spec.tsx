import { Confirm } from '@components/confirm.js'
import { render } from 'ink-testing-library'
import { delay } from 'extra-promise'

describe('Confirm', () => {
  describe('render', () => {
    test('defaultValue = true', () => {
      const fn = vi.fn()
      const { lastFrame } = render(
        <Confirm
          message='message'
          defaultValue={true}
          onSubmit={fn}
        />
      )

      expect(lastFrame()).toBe('message(Y/n):')
      expect(fn).toBeCalledTimes(0)
    })

    test('defaultValue = false', () => {
      const fn = vi.fn()
      const { lastFrame } = render(
        <Confirm
          message='message'
          defaultValue={false}
          onSubmit={fn}
        />
      )

      expect(lastFrame()).toBe('message(y/N):')
      expect(fn).toBeCalledTimes(0)
    })
  })

  describe('defaultValue', () => {
    test('defaultValue = true', async () => {
      const fn = vi.fn()
      const { stdin, lastFrame } = render(
        <Confirm
          message='message'
          defaultValue={true}
          onSubmit={fn}
        />
      )

      await delay(100)
      stdin.write('\r')
      await delay(100)

      expect(lastFrame()).toBe('message(Y/n):')
      expect(fn).toBeCalledTimes(1)
      expect(fn).toBeCalledWith(true)
    })

    test('defaultValue = false', async () => {
      const fn = vi.fn()
      const { stdin, lastFrame } = render(
        <Confirm
          message='message'
          defaultValue={false}
          onSubmit={fn}
        />
      )

      await delay(100)
      stdin.write('\r')
      await delay(100)

      expect(lastFrame()).toBe('message(y/N):')
      expect(fn).toBeCalledTimes(1)
      expect(fn).toBeCalledWith(false)
    })
  })

  describe('onSubmit', () => {
    test('yes', async () => {
      const fn = vi.fn()
      const { stdin, lastFrame } = render(
        <Confirm
          message='message'
          defaultValue={true}
          onSubmit={fn}
        />
      )

      await delay(100)
      stdin.write('yes')
      await delay(100)
      stdin.write('\r')
      await delay(100)

      expect(lastFrame()).toBe('message(Y/n): yes')
      expect(fn).toBeCalledTimes(1)
      expect(fn).toBeCalledWith(true)
    })

    test('no', async () => {
      const fn = vi.fn()
      const { stdin, lastFrame } = render(
        <Confirm
          message='message'
          defaultValue={true}
          onSubmit={fn}
        />
      )

      await delay(100)
      stdin.write('no')
      await delay(100)
      stdin.write('\r')
      await delay(100)

      expect(lastFrame()).toBe('message(Y/n): no')
      expect(fn).toBeCalledTimes(1)
      expect(fn).toBeCalledWith(false)
    })
  })
})
