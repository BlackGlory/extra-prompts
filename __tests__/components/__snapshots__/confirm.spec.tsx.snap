// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Confirm > defaultValue > defaultValue = false 1`] = `"message(y/N):"`;

exports[`Confirm > defaultValue > defaultValue = true 1`] = `"message(Y/n):"`;

exports[`Confirm > onSubmit > no 1`] = `"message(Y/n): no"`;

exports[`Confirm > onSubmit > yes 1`] = `"message(Y/n): yes"`;

exports[`Confirm > render > defaultValue = false 1`] = `"message(y/N):"`;

exports[`Confirm > render > defaultValue = true 1`] = `"message(Y/n):"`;
